spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -304,13 +306,28 @@ protected void set(InternalRow struct, int pos, Object value) {
   private static class UnionReader implements ValueReader<InternalRow> {
     private final Schema schema;
     private final ValueReader[] readers;
+    private final Map<Integer, Integer> projectedFieldIdsToIdxInReturnedRow;
+    private boolean isTagFieldProjected;
 
-    private UnionReader(Schema schema, List<ValueReader<?>> readers) {
+    private UnionReader(Schema schema, List<ValueReader<?>> readers, Type expected) {
       this.schema = schema;
       this.readers = new ValueReader[readers.size()];
       for (int i = 0; i < this.readers.length; i += 1) {
         this.readers[i] = readers.get(i);
       }
+      this.projectedFieldIdsToIdxInReturnedRow = new HashMap<>();
+      this.isTagFieldProjected = false;
+      int idxInReturnedRow = 0;
+      for (Types.NestedField icebergField : expected.asStructType().fields()) {
```

Comment:
I am not sure if the naming convention of `Iceberg` is followed in this file. Maybe we can avoid calling it `icebergField`?